aiida/manage/configuration/migrations/migrations.py

```diff
@@ -8,136 +8,282 @@
+    def upgrade(self, config: ConfigType) -> None:
+        from aiida.manage.configuration import PROFILE
 
-    if default_profiles and 'daemon' in default_profiles:
-        config['default_profile'] = default_profiles['daemon']
-    elif default_profiles and 'verdi' in default_profiles:
-        config['default_profile'] = default_profiles['verdi']
-    elif PROFILE is not None:
-        config['default_profile'] = PROFILE.name
+        default_profiles = config.pop('default_profiles', None)
 
-    return config
+        if default_profiles and 'daemon' in default_profiles:
+            config['default_profile'] = default_profiles['daemon']
+        elif default_profiles and 'verdi' in default_profiles:
+            config['default_profile'] = default_profiles['verdi']
+        elif PROFILE is not None:
+            config['default_profile'] = PROFILE.name
+
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 3
+
+    def downgrade(self, config: ConfigType) -> None:
+        if 'default_profile' in config:
+            default = config.pop('default_profile')
+            config['default_profiles'] = {'daemon': default, 'verdi': default}
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 0
 
 
-def _3_add_message_broker(config):
+class AddMessageBroker(SingleMigration):
     """Add the configuration for the message broker, which was not configurable up to now."""
-    from aiida.manage.external.rmq import BROKER_DEFAULTS
-
-    defaults = [
-        ('broker_protocol', BROKER_DEFAULTS.protocol),
-        ('broker_username', BROKER_DEFAULTS.username),
-        ('broker_password', BROKER_DEFAULTS.password),
-        ('broker_host', BROKER_DEFAULTS.host),
-        ('broker_port', BROKER_DEFAULTS.port),
-        ('broker_virtual_host', BROKER_DEFAULTS.virtual_host),
-    ]
-
-    for profile in config.get('profiles', {}).values():
-        for key, default in defaults:
-            if key not in profile:
-                profile[key] = default
+    initial = 3
+    final = 4
 
-    return config
+    def upgrade(self, config: ConfigType) -> None:
+        from aiida.manage.external.rmq import BROKER_DEFAULTS
+        defaults = [
+            ('broker_protocol', BROKER_DEFAULTS.protocol),
+            ('broker_username', BROKER_DEFAULTS.username),
+            ('broker_password', BROKER_DEFAULTS.password),
+            ('broker_host', BROKER_DEFAULTS.host),
+            ('broker_port', BROKER_DEFAULTS.port),
+            ('broker_virtual_host', BROKER_DEFAULTS.virtual_host),
+        ]
+
+        for profile in config.get('profiles', {}).values():
+            for key, default in defaults:
+                if key not in profile:
+                    profile[key] = default
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 3
 
+    def downgrade(self, config: ConfigType) -> None:
+        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = 3
```

Comment:
Again, this is not really a downgrade since it does not undo the upgrade. Could this cause issues if actually loading the downgrade config with an older version of `aiida-core`?

Reply:
Fair enough, happy to go ahead with that behavior if it will be guaranteed to work and we assume that behavior from now on